src/ResourceManager/Insights/Commands.Insights/Diagnostics/SetAzureRmDiagnosticSettingCommand.cs

```diff
@@ -99,105 +120,198 @@ public class SetAzureRmDiagnosticSettingCommand : ManagementCmdletBase
                 }
+                metricSettings.Enabled = this.Enabled.Value;
+            }
+        }
 
-                if (this.Timegrains != null)
+        private void SetSelectedCategories(ServiceDiagnosticSettingsResource properties)
+        {
+            if (!this.Enabled.HasValue)
+            {
+                throw new ArgumentException("Parameter 'Enabled' is required by 'Categories' parameter.");
+            }
+
+            foreach (string category in this.Categories)
+            {
+                LogSettings logSettings = properties.Logs.FirstOrDefault(x => string.Equals(x.Category, category, StringComparison.OrdinalIgnoreCase));
+
+                if (logSettings == null)
                 {
-                    foreach (string timegrainString in this.Timegrains)
-                    {
-                        TimeSpan timegrain = XmlConvert.ToTimeSpan(timegrainString);
-                        MetricSettings metricSettings = properties.Metrics.FirstOrDefault(x => TimeSpan.Equals(x.TimeGrain, timegrain));
-
-                        if (metricSettings == null)
-                        {
-                            throw new ArgumentException(string.Format(CultureInfo.InvariantCulture, "Metric timegrain '{0}' is not available", timegrainString));
-                        }
-                        metricSettings.Enabled = this.Enabled;
-                    }
+                    throw new ArgumentException(string.Format(CultureInfo.InvariantCulture, "Log category '{0}' is not available", category));
                 }
+
+                logSettings.Enabled = this.Enabled.Value;
             }
+        }
 
-            if (this.RetentionEnabled.HasValue)
+        private void SetAllCategoriesAndTimegrains(ServiceDiagnosticSettingsResource properties)
+        {
+            if (!this.Enabled.HasValue)
             {
-                var retentionPolicy = new RetentionPolicy
-                {
-                    Enabled = this.RetentionEnabled.Value,
-                    Days = this.RetentionInDays.Value
-                };
+                return;
+            }
 
-                if (properties.Logs != null)
+            foreach (var log in properties.Logs)
+            {
+                log.Enabled = this.Enabled.Value;
+            }
+
+            foreach (var metric in properties.Metrics)
+            {
+                metric.Enabled = this.Enabled.Value;
+            }
+        }
+
+        private void SetWorkspace(ServiceDiagnosticSettingsResource properties)
+        {
+            if (!string.IsNullOrWhiteSpace(this.WorkspaceId))
+            {
+                if (this.DisableWorkspace)
                 {
-                    foreach (LogSettings logSettings in properties.Logs)
-                    {
-                        logSettings.RetentionPolicy = retentionPolicy;
-                    }
+                    throw new ArgumentException("WorkspaceId and DisableWorkspace cannot be both present.");
```

Comment:
@vcanaa use parameter sets to ensure that these parameters are mutually exclusive